code/modules/mob/living/simple_animal/hostile/megafauna/swarmer.dm

```diff
@@ -0,0 +1,327 @@
+that's a plus I suppose.
+
+Difficulty: Special
+
+*/
+
+var/global/list/mob/living/simple_animal/hostile/swarmer/ai/AISwarmers = list()
+var/global/list/mob/living/simple_animal/hostile/swarmer/ai/AISwarmersByType = list()//AISwarmersByType[.../resource] = list(1st, 2nd, nth), AISwarmersByType[../ranged] = list(1st, 2nd, nth) etc.
+var/global/list/AISwarmerCapsByType = list(/mob/living/simple_animal/hostile/swarmer/ai/resource = 30, /mob/living/simple_animal/hostile/swarmer/ai/ranged_combat = 20, /mob/living/simple_animal/hostile/swarmer/ai/melee_combat = 10)
+
+
+//returns a type of AI swarmer that is NOT at max cap
+//type order is shuffled, to prevent bias
+/proc/GetUncappedAISwarmerType()
+	var/static/list/swarmerTypes = subtypesof(/mob/living/simple_animal/hostile/swarmer/ai)
+	LAZYINITLIST(AISwarmersByType)
+	for(var/t in shuffle(swarmerTypes))
+		var/list/amount = AISwarmersByType[t]
+		if(!amount || amount.len <  AISwarmerCapsByType[t])
+			return t
+
+
+//Total of all subtype caps
+/proc/GetTotalAISwarmerCap()
+	var/static/list/swarmerTypes = subtypesof(/mob/living/simple_animal/hostile/swarmer/ai)
+	. = 0
+	LAZYINITLIST(AISwarmersByType)
+	for(var/t in swarmerTypes)
+		. += AISwarmerCapsByType[t]
+
+
+/mob/living/simple_animal/hostile/megafauna/swarmer_swarm_beacon
+	name = "swarmer beacon"
+	desc = "That name is a bit of a mouthful, but stop paying attention to your mouth they're eating everything!"
+	icon = 'icons/mob/swarmer.dmi'
+	icon_state = "swarmer_console"
+	health = 3000
+	maxHealth = 3000 //low-ish HP because it's a passive boss, and the swarm itself is the real foe
+	medal_type = MEDAL_PREFIX
+	score_type = SWARMER_BEACON_SCORE
+	faction = list("mining", "boss", "swarmer")
+	weather_immunities = list("lava","ash")
+	stop_automated_movement = TRUE
+	wander = FALSE
+	anchored = TRUE
+	layer = BELOW_MOB_LAYER
+	AIStatus = AI_OFF
+	var/swarmer_spawn_cooldown = 0
+	var/swarmer_spawn_cooldown_amt = 150 //Deciseconds between the swarmers we spawn
+	var/call_help_cooldown = 0
+	var/call_help_cooldown_amt = 150 //Deciseconds between calling swarmers to help us when attacked
+	var/static/list/swarmer_caps
+
+
+/mob/living/simple_animal/hostile/megafauna/swarmer_swarm_beacon/New()
+	..()
+	swarmer_caps = AISwarmerCapsByType //for admin-edits
+	internal = new/obj/item/device/gps/internal/swarmer_beacon(src)
+	for(var/ddir in cardinal)
+		new /obj/structure/swarmer/blockade (get_step(src, ddir))
+		var/mob/living/simple_animal/hostile/swarmer/ai/resource/R = new(loc)
+		step(R, ddir) //Step the swarmers, instead of spawning them there, incase the turf is solid
+
+
+/mob/living/simple_animal/hostile/megafauna/swarmer_swarm_beacon/Life()
+	. = ..()
+	if(.)
+		var/createtype = GetUncappedAISwarmerType()
+		if(createtype && world.time > swarmer_spawn_cooldown && AISwarmers.len < (GetTotalAISwarmerCap()*0.5))
+			swarmer_spawn_cooldown = world.time + swarmer_spawn_cooldown_amt
+			new createtype(loc)
```

Comment:
can you use https://github.com/tgstation/tgstation/blob/master/code/__HELPERS/mobs.dm#L333 spawn_atom_to_turf or otherwise check that the location is valid and correct.